### ISSUE TITLE
Location not updating when the app is in background.

### ISSUE THREAD

**Issue by LukePenkava** [transistorsoft/react-native-background-geolocation#857]
## Your Environment
* Plugin version: 3.2.2
* Platform: iOS 
* Device manufacturer / model: iPhone6
* React Native version (`react-native -v`): 0.59.10
* Plugin config
```javascript <!-- Syntax-highlighting:  paste your code below -->

desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
distanceFilter: 1,                                                
stopTimeout: 1,
debug: true,
logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
stopOnTerminate: false
 startOnBoot: true,       
url: `https://testProject1234.firebaseio.com/Users/${user_key}/location.json`,
batchSync: false,      
autoSync: true,       
headers: {   },
params: {    }
```

## Expected Behavior
Send location data to the database when the app is in the background

## Actual Behavior
Whenever the app goes to background no data are sent, whether moving or being stationary. I have tested this both on simulator and real device. With iOS simulator i am using debugging like city run etc. When i am in the app, i can see the data being synced nicely and all is good. As soon as the app goes to background it immediately stops. Exact same behaviour happens on iPhone6. I have seen some sporadic update, like 15 minutes later one location update will be sent randomly, but not reliably.

**Edit:** The notifications also immediately stop after the app is put to background. Unlike the youtube demo you made, where you have notifications when app is in foreground and also when it is in background. For me the notifications immediately stop when the app goes to background.

## Steps to Reproduce
I have followed the template and did the exact same installation. 

## Context
I am modifying the url with userKey parameter to be able to store the location data directly to his profile. I have read bunch of other issues here and was not able to figure out why its not working properly.

I have also question concerning  onLocation(location) function. Is it possible to send custom parameter there ? I have tried to send the location data to my db from this function and it seems to be working ( again, nothing works when the app is in the background ). I was basicly able to do this:

```
firebase.database().ref(`Users/userkey123456/location`)
            .push({
                lat: location.coords.latitude
            })
```

but i would need to be able to send there the user key like this

```
onLocation(location, userKey) {
 firebase.....
}
```

Thank you

**Edit:**
I have also setup hearbeat on iOS, that works well in background in simulator as long as simulator is not moving, then it stops working. 
 Ie simulator has set location, app is installed, data are sent to the db, all good. App is put to background, data are still sent to the db, all good. I setup city run or anything like that for the simulator and the data are not being sent anymore. Making the simulator stationary again does not help to receive the data.

 Also, even if i got this working, which is honestly all i need, just some way of updating the location data in the db once in a while, i am still not sure how to use the userKey for the db i have stored in the Component state. Based on what i read in your docs, i cant put this code somewhere deep in the UI, so it can get triggered by iOS, but in ComponentWillMount, i dont have access to the userKey data yet ( this thing  firebase.database().ref(`Users/-Lpsuyd7mChBYXF1zSgq/location`) )

```
  BackgroundGeolocation.ready({
            reset: true,
            preventSuspend: true,       
            desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
            distanceFilter: 5,        
            stopTimeout: 1,
            heartbeatInterval: 20,           
            debug: true, 
            logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
            stopOnTerminate: false,  
            startOnBoot: true
        }, (state) => {
            if (!state.enabled) {            
                BackgroundGeolocation.start(function () {
                    console.log("- Start success");
                });
            }
        });
        
        BackgroundGeolocation.onHeartbeat((event) => {    
            BackgroundGeolocation.getCurrentPosition({
                samples: 1,
                persist: true
            }).then((location) => {              
                firebase.database().ref(`Users/-Lpsuyd7mChBYXF1zSgq/location`)
                    .push({
                        lat: location.coords.latitude
                    })
                    .catch(error => console.log(error))
            });
        })
```

**Comment by LukePenkava** [comment]
I have implemented this solution and it seems to work well so far in the background, not sure what is the major change there honestly, but its working with the app in the background, locked phone and it uploads data to firebase database. So perhaps it will help someone else in the future.

```
componentWillMount() {
        BackgroundGeolocation.onLocation(this.onLocation, this.onError);
       
        BackgroundGeolocation.ready({
            reset: true,
            preventSuspend: true,
            desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
            distanceFilter: 100,
            stopTimeout: 1,
            debug: false,
            logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
            stopOnTerminate: false,  
            startOnBoot: true,             
        }, (state) => {
            if (!state.enabled) {               
                BackgroundGeolocation.start(function () {
                    console.log("- Start success");
                });
            }
        });
    
        this.setState({
            willMountTrue: true
        }, () => this.requestLocationPermission());
    }

    componentWillUnmount() {
        BackgroundGeolocation.removeListeners();
    }

    onLocation(location) {
        if (location.coords.speed < 10) {
            firebase.database().ref('Users')
                .orderByChild('email')
                .equalTo(firebase.auth().currentUser.email)
                .once('value')
                .then(snapshot => {
                    if (snapshot.val()) {
                        snapshot.forEach(user => {
                            firebase.database().ref(`Users/${user.key}/location`)
                                .push({
                                    lat: location.coords.latitude,
                                    long: location.coords.longitude,
                                    speed: location.coords.speed,
                                    timestamp: location.timestamp
                                })
                                .catch(error => console.log(error))
                        })
                    }
                })
        }
    }
```

**Comment by christocracy** [comment]
That query seems awfully inefficient to execute with each recorded location.

Can't you get a reference to that Firebase user just once, when your app launches and store that `$user.key} into your State?

Also, be are of `location.sample == true` in `#onLocation` -- you don't want to persist those locations.  See the docs for `#onLocation` event for what `location.sample` is.

**Comment by LukePenkava** [comment]
@christocracy I have implemented it with storing userKey in the state, but i am getting sometimes undefined object. I would prefer to store it in the state, but i had constant issues with it, this is only way i was able to get some reliable results unfortunately. I need the location update only once in about 1000 metres when stationary, so it should be hopefull ok.

Concerning the location.sample. I read the docs, if i got it right, it means that i want to ignore the location update if its location.sample == true ? So can i do something like this :

```
if(location.sample == false) {
 //update firebase db
}
```

I have noticed from my field test, that i am getting a lot of location updates around the same position, i am guessing the location.sample has something to do with it ? Thanks
